### PR TITLE
feat: return error code upon failure

### PR DESCRIPTION
**Pull request by kririae**
* fix: fix hangling with `-i` only
* feat: don't terminate processing in case of failure

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -53,33 +58,62 @@ fn main() {
                             continue;
                         };
                         let res = Typstyle::new_with_content(content, width).pretty_print();
-                        std::fs::write(entry.path(), res).unwrap();
-                        format_count += 1;
+
+                        // `FormatAll` must be done in place without failing in the middle
+                        match std::fs::write(entry.path(), res).with_context(|| {
+                            format!("Failed to write to the file {:#?}", entry.path())
```
Comment (Myriad-Dreamin) [inline src/main.rs:64]: Error messages shouldn't be capitalized.

File: src/main.rs [hunk 2]
```diff
@@ -53,33 +58,62 @@ fn main() {
...
+                        }) {
+                            Ok(_) => format_count += 1,
+                            Err(e) => {
+                                eprintln!("{}", e);
```
Comment (Myriad-Dreamin) [inline src/main.rs:68]: Format arguments should be inline whenever possible.
```rs
eprintln!("{e}");
```

File: src/main.rs [hunk 3]
```diff
@@ -53,33 +58,62 @@ fn main() {
...
+                                err_files.push(entry.path().to_path_buf());
+                            }
+                        }
                     }
                 }
-                println!("Formatted {} files", format_count);
+
+                println!("Successfully formatted {} files", format_count);
```
Comment (Myriad-Dreamin) [inline src/main.rs:75]: Should we output all messages to `stderr`?

File: src/main.rs [hunk 4]
```diff
@@ -53,33 +58,62 @@ fn main() {
...
         for file in &args.input {
-            format(Some(file), &args);
+            if let Err(e) = format(Some(file), &args) {
+                eprintln!("{}", e);
+                err_files.push(file.clone());
+            }
+        }
+
+        if !err_files.is_empty() {
+            bail!("Failed to process the following file(s): {:#?}", err_files);
```
Comment (Myriad-Dreamin) [inline src/main.rs:98]: An error message shouldn't contain newlines.

File: src/main.rs [hunk 5]
```diff
@@ -98,11 +142,14 @@ fn format(input: Option<&PathBuf>, args: &CliArguments) {
     };
     if *inplace {
         if let Some(input) = input {
-            std::fs::write(input, res).unwrap();
+            std::fs::write(input, res)
+                .with_context(|| format!("failed to write to the file {:#?}", input))?;
         } else {
-            panic!("Cannot use inplace formatting with stdin");
+            panic!("cannot perform in-place formatting without at least a file presented");
```
Comment (seven-mile) [inline src/main.rs:148]: Nit:

```suggestion
            panic!("cannot perform in-place formatting without at least one file being presented");
```

**Review by seven-mile — APPROVED** [review]

File: src/main.rs (hunk 5 above)
Comment (kririae) [reply src/main.rs]: Commited.

File: src/main.rs [hunk 6]
```diff
@@ -12,13 +14,16 @@ use crate::cli::CliArguments;
 
 mod cli;
 
-fn get_input(input: Option<&PathBuf>) -> String {
+fn get_input(input: Option<&PathBuf>) -> Result<String> {
     match input {
-        Some(path) => std::fs::read_to_string(path).unwrap(),
+        Some(path) => std::fs::read_to_string(path)
+            .with_context(|| format!("failed to read the file {:#?} into string", path)),
```
Comment (Enter-tainer) [inline src/main.rs:20]:
```suggestion
            .with_context(|| format!("failed to read file {:#?}", path)),
```

BTW, what does the output looks like? `{:#?}` looks strange here

File: src/main.rs [hunk 7]
```diff
@@ -53,33 +68,62 @@ fn main() {
                             continue;
                         };
                         let res = Typstyle::new_with_content(content, width).pretty_print();
-                        std::fs::write(entry.path(), res).unwrap();
-                        format_count += 1;
+
+                        // `FormatAll` must be done in place without failing in the middle
+                        match std::fs::write(entry.path(), res).with_context(|| {
+                            format!("failed to write to the file {:#?}", entry.path())
```
Comment (Enter-tainer) [inline src/main.rs:74]:
```suggestion
                            format!("failed to overwrite file {:#?}", entry.path())
```

same as above

File: src/main.rs [hunk 8]
```diff
@@ -31,8 +36,17 @@ fn is_hidden(entry: &DirEntry) -> bool {
         .unwrap_or(false)
 }
 
-fn main() {
+fn main() -> CliResults {
     let args = CliArguments::parse();
+    if let Err(e) = execute(args) {
+        eprintln!("typstyle: {e}");
```
Comment (Enter-tainer) [inline src/main.rs:42]: is `typstyle: ` needed?

File: src/main.rs [hunk 9]
```diff
@@ -98,11 +142,14 @@ fn format(input: Option<&PathBuf>, args: &CliArguments) {
     };
     if *inplace {
         if let Some(input) = input {
-            std::fs::write(input, res).unwrap();
+            std::fs::write(input, res)
+                .with_context(|| format!("failed to write to the file {:#?}", input))?;
         } else {
-            panic!("Cannot use inplace formatting with stdin");
+            panic!("cannot perform in-place formatting without at least one file being presented");
```
Comment (Enter-tainer) [inline src/main.rs:148]: is it possible to switch to `bail!` here?

File: src/main.rs [hunk 10]
```diff
@@ -1,24 +1,28 @@
...
+use crate::cli::{CliArguments, CliResults};
 
 mod cli;
 
-fn get_input(input: Option<&PathBuf>) -> String {
+fn get_input(input: Option<&PathBuf>) -> Result<String> {
     match input {
-        Some(path) => std::fs::read_to_string(path).unwrap(),
+        Some(path) => std::fs::read_to_string(path)
+            .with_context(|| format!("failed to read the file {:#?} into string", path)),
```
Comment (Enter-tainer) [inline src/main.rs:19]:
```suggestion
            .with_context(|| format!("failed to read {:#?}", path)),
```

File: src/main.rs [hunk 11]
```diff
@@ -53,33 +67,65 @@ fn main() {
                             continue;
                         };
                         let res = Typstyle::new_with_content(content, width).pretty_print();
-                        std::fs::write(entry.path(), res).unwrap();
-                        format_count += 1;
+
+                        // `FormatAll` must be done in place without failing in the middle
+                        match std::fs::write(entry.path(), res).with_context(|| {
+                            format!(
+                                "failed to write to the file {path}",
```
Comment (Enter-tainer) [inline src/main.rs:74]:
```suggestion
                                "failed to overwrite {path}",
```

File: src/main.rs [hunk 12]
```diff
@@ -53,33 +67,65 @@ fn main() {
...
     let CliArguments {
         column: line_width,
         ast,
         pretty_doc,
         inplace,
         ..
     } = args;
-    let content = get_input(input);
+    if *inplace && input.is_none() {
+        bail!("cannot use inplace format with only stdin");
```
Comment (Enter-tainer) [inline src/main.rs:126]:
```suggestion
        bail!("cannot perform in-place formatting without at least one file being presented");
```

File: src/main.rs [hunk 13]
```diff
@@ -98,11 +144,19 @@ fn format(input: Option<&PathBuf>, args: &CliArguments) {
...
+            std::fs::write(input, res).with_context(|| {
+                format!(
+                    "failed to write to the file {file}",
+                    file = input.display().to_string()
+                )
+            })?;
         } else {
-            panic!("Cannot use inplace formatting with stdin");
+            // This branch should never be reached
+            panic!("cannot perform in-place formatting without at least one file being presented");
```
Comment (Enter-tainer) [inline src/main.rs:155]:
```suggestion
            unreachable!("cannot perform in-place formatting without at least one file being presented");
```

**Pull request merged**